[SDBELGA-126] Integrate Video Server to Superdesk Client

idsvn opened the pull request:
This PR based on #3101 but integrate as an extensions and fixes some issues as stated in https://github.com/superdesk/superdesk-client-core/pull/3101#pullrequestreview-287506125
For back-end, we are waiting for superdesk/video-server#15 and superdesk/superdesk-core#1668

New changes:
+ Keep crop position when rotating
+ Video will scale (up / down) to fit into container but do not exceed max video size to avoid breaking video quality
+ Change [confirm dialog](https://user-images.githubusercontent.com/45254285/69032791-e3bfb380-0a0f-11ea-9850-9d9ec831b755.png) on preview thumbnail

Issues:
+ Chrome [add 10 to position](https://user-images.githubusercontent.com/45254285/69046209-057a6400-0a2b-11ea-8513-d0aa1cc7028d.png), cause ReactImageCrop display area mismatch video, but Firefox works fine so we [calculate margin top](https://github.com/idsvn/superdesk-client-core/blob/cbb9b3c568b54987e73340c995db071f10520841/scripts/extensions/videoEditor/src/VideoEditor.tsx#L255) on render to avoid this
+ For optional chaining, there is a bug in Typescript so if property nullable, we must write `!?` to avoid compiler warning
+ How to update article outside extension? We need to reload video and thumbnails, previously, we use `superdesk.entities.article.update` but at the moment, it no longer works, should we add random param to `renditions` thumbnail url to *force* reload thumbnail?

tomaskikutis commented:
Can you fix the lint issues? If you don't know how to fix the TypeScript error`Property 'componentRef' does not exist on type 'ReactCrop'.` you can leave it and I'll take a look, but please fix the style issues.

You can check the style issues by running `npm run lint` in the root of the repository(you can remove `tsc -p scripts --noEmit ` step in order to see style issues only).

idsvn commented:
We fixed all linting issues from `lint` command for all files in this PR, it seems `lint-fix` command has a bug cause `authoring.spec.ts` look [like this](https://user-images.githubusercontent.com/45254285/69124466-8a6e8780-0ad6-11ea-8321-f74731ea0e65.png) so we don't touch these

May you have a look at `componentRef` and Travis, we only apply temporary change to `current?.['componentRef']` to suppress warning, also we build normally in local but travis build failed.

ride90 commented:
Hi @idsvn , good job 👍  
I've tested and it works for me. The only issue is that after a video was edited ("done" button was clicked) player still has an old video loaded.

tomaskikutis commented:
> The only issue is that after a video was edited ("done" button was clicked) player still has an old video loaded.

That's a known issue, I'll be taking a look at it.

tomaskikutis commented:
Hi @idsvn, just letting you know that I've pushed a few fixes to your branch. There are issues which are quite simple and would be harder to explain than to fix.

I haven't finished the review yet so you don't need to fix anything yet. I'll submit the review on GitHub when I'm done.

idsvn commented on scripts/extensions/videoEditor/src/VideoEditorThumbnail.tsx:
```diff
@@ -0,0 +1,361 @@
+    type: 'capture' | 'upload' | '';
+    value: number | File; // capture positon or uploaded File
+    // most recent updated thumbnail, for re-drawing thumbnail when reset changes
+    thumbnail: string;
```
We update article on saving thumbnail so it's redundant now, will remove it.
Previously, if video is just uploaded, article thumbnail is null even after updating thumbnail (because we think article does not need to be updated, but it is, because there is etag comparison)

idsvn commented on scripts/extensions/videoEditor/src/VideoEditor.tsx:
```diff
@@ -0,0 +1,683 @@
+    componentDidMount() {
+        const canvas = document.createElement('canvas');
+
+        canvas.width = 2000;
```
Whichever value is fine, as long as it's big enough to cover entire video. This is maximum area of React Crop, we will resize it to match video later, but it can't exceed init image size.
We will add a small note to it to make it more clear.

idsvn commented on scripts/extensions/videoEditor/src/VideoEditor.tsx:
```diff
@@ -0,0 +1,688 @@
+
+        const {gettext} = this.props.superdesk.localization;
+
+        if (body.crop === '0,0,0,0') {
```
This change is based on https://github.com/superdesk/video-server/issues/17

idsvn commented on scripts/extensions/videoEditor/src/VideoEditor.tsx:
```diff
@@ -0,0 +1,688 @@
+    thumbnails: Array<IThumbnail>;
+    loading: boolean;
+    loadingText: string;
+    scale: number;
```
This is used to scale video to fit into container (but not exceed maximum video size) to avoid video [too small](https://user-images.githubusercontent.com/2076953/64971798-817bf280-d8a8-11e9-99db-8d203527e4c9.png) or overflow

tomaskikutis replied:
Couldn't it work CSS only?

idsvn replied:
We have tried it before but couldn't make it to work.
We need to scale video down (or up) when applied rotate transformation. If scale video up, we must ensure that it's not exceeded video size to avoid breaking video images. Do you have any suggestions?

tomaskikutis replied:
I've looked it up and understand it now. When you apply a CSS transform, the parent element doesn't take up the dimensions of the transformed element inside.

Keep it as it is then.

idsvn commented on scripts/extensions/videoEditor/src/VideoTimeline/VideoTimeline.tsx:
```diff
@@ -0,0 +1,251 @@
+}
+interface IState {
+    currentTime: number;
+    thumnailsRender: Array<IThumbnail>;
```
Because, we don't want capture thumbnails (in server) every "UI Video editor open". 
To do this, we will capture 50 (can change in settings) pictures and store them in server first. So when video editor open, we don't need capture again, we only choose a required number of thumbnails (`thumnailsRender`) in these 50 pics. 
That means `thumbnails` is data of existed 50 pictures in server. and `thumnailsRender` is a required number of thumbnails that will be rendered in UI, they are base on width of timeline (maybe 10-15 pictures) for filling up all timeline control.

idsvn commented on scripts/extensions/videoEditor/src/VideoEditorThumbnail.tsx:
```diff
@@ -0,0 +1,361 @@
+
+    componentDidMount() {
+        if (this.state.thumbnail) {
+            const thumbnail = this.state.thumbnail + `?t=${Math.random()}`;
```
It can, but `Cache-Control` needs to be `no-store` for it to work, this mean thumbnail will be re-loaded every time user click Cancel. 
While checking this, it seems `thumbnail` state in above comment is needed so that thumbnail will not need to reload on canceling. 
Should we leave this for server to handle?

tomaskikutis replied:
> It can, but Cache-Control needs to be no-store for it to work, this mean thumbnail will be re-loaded every time user click Cancel.

You should be able to apply different caching policies based on the endpoint. If it's hard to do, remove caching for all video editor endpoints. Thumbnails don't require much data and I don't think the use case of closing and reopening the same video would be that common.

idsvn replied:
Sorry if not make clear, by reload thumbnail, we mean capturing a preview thumbnail then click cancel, if there is no caching in place, it will reload each time user click capture then cancel.
But we do agree, we should not worry about reloading thumbnail, if thumbnail size is too big, it's server responsibility to optimize it

idsvn commented on scripts/extensions/videoEditor/src/VideoTimeline/ListThumbnails.tsx:
```diff
@@ -0,0 +1,31 @@
+            <div className={`${getClass('frames')} ${getClass('frames--thumbs')}`}>
+                <div className={getClass('frames__inner')}>
+                    {this.props.thumbnails.map((item: IThumbnail, index: number) => (
+                        <video
```
We use `video` element. Because it support get thumbnails directly in video stream(via time). But it very slow. it will faster if we use `blob` (but it must pre-download file to browser)
So we change to use solution: pre-capture 50 pics in server. We keep video element because we want to find another way to make it faster in future.
But we think we should change to `img` element for now. Maybe we will update again to video later if we find any way better.

idsvn replied:
We use `video` element. Because it support get thumbnails directly in video stream(via time). But it very slow. it will faster if we use `blob` (but it must pre-download file to browser)
So we change to use solution: pre-capture 50 pics in server. We keep video element because we want to find another way to make it faster in future.
But we think we should change to `img` element for now. Maybe we will update again to video later if we find any way better.

tomaskikutis replied:
Thanks for explaining. Generating thumbnails on the server is technically simpler and will probably always be faster since you can do it as soon as the video is uploaded while if you wanted to do it on the client you have to download parts of the video which probably doesn't work in parallel either.

idsvn commented:
Hi @tomaskikutis, the issue:

> Changing video trim boundaries should not change the current video position. Let's say a video is 20 seconds long and is paused at 10s. If I want to trim 2 seconds from the start or the end of the video, it should stay paused at 10s after setting the trim boundaries.

Yes, we understand your idea. We did it before. But when we testing, we see it have little difficult for user when they cutting video. Because, user will attend to start and end time when they trim video. That mean when they move the start bar, they will attend to position of start in video, so they want to see the current video in that time for recognizing trimming time is exactly. If we don't move position of current video, they have to click more one time to start position for checking it, it have little difficult for user. So IMO, we should move current time to start or end when user want to trim video.
What do you think about this?
Note: Reference https://online-video-cutter.com/

idsvn commented on scripts/extensions/videoEditor/src/VideoEditor.tsx:
```diff
@@ -0,0 +1,688 @@
+    loadingText: string;
+    scale: number;
+    videoSrc: string;
+    article: IArticle;
```
Because we can't update article via `props.onArticleUpdate`, it seems opening `videoEditor` cause article to lock so it logs error *Article is locked and can't be updated via HTTP directly.The updates will be added to existing diff in article-edit view instead.*, lead to mismatch etag on next api call

idsvn commented on scripts/extensions/videoEditor/src/VideoTimeline/VideoTimeline.tsx:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    componentDidMount() {
+        // call tick every 100ms to update current time state
```
We've used `timeupdate` event. However, the loop time of it is 250ms, it's little bit long, we cannot setting it, it base on system, [references](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/timeupdate_event). So we have a issue, It make Video is like [delay](https://drive.google.com/open?id=1a1fxqz3QHwpU0Zdj167rUJlTW_rbcpMe).
Should we keep the interval or use `timeupdate`?

tomaskikutis replied:
Do you mean that the timeline progress indication is delayed a bit as the video is playing? I could see that too. Keep the interval solution then. Add a comment with explanation which tick is used instead of `timeupdate`

idsvn commented on scripts/extensions/videoEditor/src/VideoEditor.tsx:
```diff
@@ -0,0 +1,688 @@
+        this.reactCropMarginDelta = 0;
+        this.setState({cropEnabled: !this.state.cropEnabled, crop: crop}, () => {
+            this.checkIsDirty();
+            // chrome adds 1rem extra (ghost) margin to ReactCrop cause crop area and video mismatch
```
We think it's Chrome's fault, as you can see, [Firefox](https://user-images.githubusercontent.com/45254285/70311265-e6aa0900-1843-11ea-8fd9-552da267ba7f.png) does not need subtract margin for Video and React Crop to match but [Chrome does](https://user-images.githubusercontent.com/45254285/70311295-f9244280-1843-11ea-8951-5fd52ed4e704.png) (because it added [some extra position](https://user-images.githubusercontent.com/45254285/70312301-f0cd0700-1845-11ea-8a5d-1231a9967829.png)), if we remove margin for both, while it can reduce complex calculation, the [video is right next](https://user-images.githubusercontent.com/45254285/70311605-88315a80-1844-11ea-892a-4a7623791e29.png) to confirm bar, make it look a bit unnatural.

Should we remove margin for both video and react crop or look for other simpler solution?

ride90 commented on scripts/extensions/videoEditor/src/VideoEditor.tsx:
```diff
@@ -0,0 +1,713 @@
+    loadTimelineThumbnails() {
+        this.intervalThumbnails = window.setInterval(() => {
+            this.props.superdesk.dataApi
+                .findOne<IArticle>('video_edit', this.state.article._id + `?t=${Math.random()}`)
```
What the purpose of getting an entire `project` here? 
`.findOne<IArticle>('video_edit', this.state.article._id + `?t=${Math.random()}`)`

If we want to load thumbnails why not to call only this:
`.findOne('video_edit', this.state.article._id + `?action=timeline&t=${Math.random()}`)` ?
If there are no thumbnails video server will run a task to create them and return 202 status, if the response status is 409, it means that the task is in the process, if it's 200 we get a list of thumbnails.

idsvn commented on scripts/extensions/videoEditor/src/VideoTimeline/VideoTimeline.tsx:
```diff
@@ -0,0 +1,253 @@
+        let time = this.setVideoCurrentTime(e.clientX);
+
+        if (time < this.props.trim.start) {
+            this.props.onTrim(time, this.props.trim.end, false);
```
We have update this. This logic use for UX, when user click outer of trim, it will set trim and current time at this position.
We call `onTrim` to update Trim in `videoEditor`, and `trimUpdate` for update state trim in `videoTimeline`.
![video-_2_](https://user-images.githubusercontent.com/45254285/71164284-53c89000-2281-11ea-9303-fa7ef9bb39e2.gif)

tomaskikutis replied:
got it, thanks

idsvn commented on scripts/extensions/videoEditor/src/VideoTimeline/VideoTimeline.tsx:
```diff
@@ -0,0 +1,255 @@
+        const numberThumbnails = Math.floor(this.controlbar.current!?.offsetWidth / widthPic);
+        const duration = video?.duration ?? 0;
+        let thumbnailsRender: Array<IThumbnail> = [];
+        const per_delta_image =
```
We choose a required number of thumbnails (thumbnailsRender) in 50 pics(on server). So we will calculate average index of pictures that fit with each of time in timelinebar.
average_index = number_thumbnails_render/ number_pics_on_server
After that, we use average_index to get congruous picture that will show on timeline.

idsvn commented on scripts/extensions/videoEditor/src/VideoEditor.tsx:
```diff
@@ -0,0 +1,683 @@
+
+    checkIsDirty = () => {
+        const state = pick(this.state, ['crop', 'trim', 'degree', 'quality']);
+        // ignore trim.end as initState don't load video duration due to videoRef can be null when component did mount
```
By non-`null` video element, do you mean we only render `VideoEditor` component only when its `Wrapper` finished loading and after we got video url? Or should we use callback ref to load video duration during componentDidMount? Could you elaborate more? Thanks
We have tried to use `Wrapper` solution as you suggested, but at the moment, `VideoEditorHeader` is depending on `VideoEditor`, so if we don't render `VideoEditor`, there is nothing left to show, but if we move `Header` outside, we need to pass props around. And actually, we can refactor [handleCheckingVideo](https://github.com/superdesk/superdesk-client-core/pull/3218/files#diff-6a26d285a3bde2bda5233a909812dbe5R147) to reuse it on `handleSave`

tomaskikutis replied:
I was looking at the code again and I think you don't need a wrapper after all. Sorry for giving a wrong direction.

What I would like do is to ensure that video element is not null as early as possible so we don't need the checks later. For example, it would help accepting a non-null video element as an argument in methods instead of using a possibly null instance value. See https://github.com/idsvn/superdesk-client-core/commit/632a9e5eb422e2705d794da440723f58c36751de

tomaskikutis replied:
You won't be able to avoid all null checks, but use this patterns where you can.

idsvn replied:
Got it, thank you, we will update the code accordingly

idsvn commented on scripts/extensions/videoEditor/src/VideoTimeline/VideoTimeline.tsx:
```diff
@@ -0,0 +1,255 @@
+    componentDidMount() {
+        // call tick every 100ms to update current time state
+        this.intervalTimer = window.setInterval(this.tick, 100);
+        document.addEventListener('dragover', this.handledragover);
```
We use it for fixing issue `cannot drag the left-right bar of timeline in FireFox`. 
Because in Firefox, `event.clientX`(mouse position) cannot get when is dragging (drag event), it's only get in `dragover` event of window. So we use `dragover` event window to save positionX( mouse x position) and use it when is dragging.

tomaskikutis replied:
Got it. Can you add comments in the code when implementing such workarounds like browser APIs not working well like in this case?

You still need to remove the event listener in `componentWillUnmount` lifecycle hook.

idsvn commented on scripts/extensions/videoEditor/src/interfaces.ts:
```diff
@@ -0,0 +1,28 @@
+import {Crop} from 'react-image-crop';
+
+export interface ICrop extends Crop {
+    scale?: number;
```
It use for calculate the drop size when we have do both action drop and rotate. When we rotate the video, it will scale to fit with window video-editor. That is the scale value of video (ICrop.scale), we store and use it for calculate the drop size after rotated.

idsvn commented on scripts/extensions/videoEditor/src/VideoEditor.tsx:
```diff
@@ -0,0 +1,688 @@
+            dataApi
+                .create('video_edit', {
+                    edit: body,
+                    item: this.state.article,
```
Good point, we only need `_id`, `media` (video id in this case) and `renditions`, we will update it accordingly

idsvn commented:
Hi @tomaskikutis, we have finished updating, here are summary:

Highlight some fixes based on your review:
- Truncate long video name
- Make server return consistent error response so error message will no longer display json ([commit](https://github.com/superdesk/video-server/commit/525b83c857ac58f89a165dc8d602c1fae5f8a723) and [commit](https://github.com/superdesk/superdesk-client-core/pull/3218/commits/a1f62c317ec532fe9545066a371bc5dd129b53f3))
- Use `PureComponent` and remove `any` type
- Move video changes to `Transformation` state
- Calculate scale video based on dynamic value from `ref` instead of hard-code value
- Return `isDirty` instead of syncing it to state
- Remove `data-content` usage
- We removed caching hack for GET video status **only**, see **Known issues** for more info, here is our nginx configuration (`more_set_headers` is from `nginx-extras`):
```
        location ~ /api/video_edit/[0-9a-zA-Z\.\:\-]+ {
            proxy_pass http://127.0.0.1:5000;
            more_set_headers "Cache-Control: no-store";
        }
```

Changes:
- **Remove** ReactContext, because we only have at most 3 nested child so typed ReactContext is not much better than passing props ([commit](https://github.com/superdesk/superdesk-client-core/pull/3218/commits/74be098eeb8d4d09de7c5dda5ca3b47a90c2226e))
- Fix timeline bar overlap (left bar position > right bar position), we use hard-code 2 minutes here, we don't need to change if we need longer (e.g. 3m) because server will raise error to let user know
- Fix incorrect calculation for expected video value when applying multiple transform
- Remove scale and value from crop state ([commit](https://github.com/superdesk/superdesk-client-core/pull/3218/commits/3ce8b4ce80b69f38e0964d157d619212be09c182))
- Use `durationchange` instead of `loadeddata` event, we think it make more sense because we only need to update video duration so don't need to wait until `loadeddata` event
- Fix Interval polling does not clear in some cases ([commit](https://github.com/superdesk/superdesk-client-core/pull/3218/commits/a0a07743cf1a9ee6eab618264b633a3c72539bf4) and [commit](https://github.com/superdesk/superdesk-client-core/pull/3218/commits/aa7edd19084eb8a018af916c5df4fbdc52851a5c))

Known issues:
- Sometime video will not apply transformations, it is because we pass odd scale value, causing video server `ffmpeg` fail, you can use our `safari branch` on video-server-app: `git+git://github.com/idsvn/video-server.git@safari#egg=videoserver` for the time being, we will patch it at once when this done
- Firefox will not reload image even when setting `no-store` for `Cache-Control` header, it seems it only respects `Cache-Control` header when doing full page reload, you can easily reproduce this by Load image with `no-store` header > replace it > re-add it to page ([example](https://user-images.githubusercontent.com/45254285/85268646-fb2cb680-b4a0-11ea-99e9-a698ebea07d9.gif)), so we have to use caching hack for image (preview, timeline)
- If you drag timeline bar too fast, [it will not stop at 2m exactly](https://user-images.githubusercontent.com/45254285/85269429-10eeab80-b4a2-11ea-8c9e-c84bab3328e7.gif) as expected, but we think this is not a big issue
- If you cut video to 2m, you can still drag to cut it down more, although [it still display 2m](https://user-images.githubusercontent.com/45254285/85269744-85294f00-b4a2-11ea-9cf7-3a88bec6c66e.gif), it is because we decimal value for trim
- Could not update article on exit due to locking, we think we could use event trigger to save article (for autosave Coming Up) for this, we will try if it's worked when it's done
- We could not make `flex` work to replace `max-width: calc(100vw - 6rem - 16rem * 16/9)` in css

tomaskikutis commented:
> Could not update article on exit due to locking, we think we could use event trigger to save article (for autosave Coming Up) for this, we will try if it's worked when it's done

There was an issue in the core. After `src` is updated, the video element has to be reloaded. I've fixed that in 
c792d4a.

There are a few more issues:
* Looks like `IArticle` type is used in places where another type should be used. For example `findOne<IArticle>('video_edit')`. `video_edit` endpoint doesn't return `IArticle` entity. Check other usages too.
* [Thumbnails](https://user-images.githubusercontent.com/2076953/85521942-45c74380-b605-11ea-8bac-2779b7e653fe.png) don't match the time of the video.

I'm also not sure whether video-editing should cancellable or not. For example if you change the headline or another field in the article, you can discard changes. Should users be able to discard changes after clicking "done" after editing the video? If not, you will have to revert my latest commit - 558fb26

tomaskikutis commented:
Also, in 5064af1 commit, I only pass `renditions` for patching. Add more fields if you need to, but keep in mind that the second argument to the `patch` function must be a diff. You can't pass the entire `this.state.article` object for example - you would need to compute a diff first.

idsvn commented:
> IArticle type is used in places where another type should be used

Good point, we don't save it in db so it's better to use `IVideoProject` and remove `project` it from `IArticle`.

> Thumbnails don't match the time of the video

Thumbnails only match the time of the video to some extent, the longer the video is, the more inaccurate it is

> Should users be able to discard changes after clicking "done" after editing the video?

At the moment, there is no way to cancel an ongoing editing operation

> Also, in 5064af1 commit, I only pass renditions for patching

Update only renditions is enough. If the client could patch renditions, should we include `baseImage` and `viewImage` too? Those two still use the old thumbnail and of course we should check the size first. Maybe better to do it in `superdesk-core`? 

IMO, we should have general function to handle these to avoid having to keep it in sync, just for future convenient when we need to add new field

tomaskikutis commented:
> Thumbnails only match the time of the video to some extent, the longer the video is, the more inaccurate it is

I understand that thumbnails can't be 100% accurate, but the problem that I am referring to is that the image that is displayed as a thumbnail NEVER appears in the time range that the thumbnail is representing.

If you look at the [image I linked](https://user-images.githubusercontent.com/2076953/85521942-45c74380-b605-11ea-8bac-2779b7e653fe.png) in a previous comment, there are 2 types of scenes - one is where the camera is pointing to land and another to sky. At `00:00:08.4` the land scene is already over and is not shown again in the rest of the video, but the next thumbnail shows a picture of a land scene anyway. Here's the [entire video](https://drive.google.com/file/d/1SvBvhsSGn3eL9vMFxymsy12HZSJoGD1s/view?usp=sharing).

> At the moment, there is no way to cancel an ongoing editing operation

I meant to ask a different thing. If the editing has already finished and you click "done", you get back to the article editing view. In article editing view, there's [a button](https://user-images.githubusercontent.com/2076953/85856778-e27e1280-b7b8-11ea-90d3-7b01e75d8d87.png) to cancel all changes done since last time "save" was clicked. In video editing case I assume it shouldn't be allowed, because after the video was edited, there is no way to undo it or is there? If not, you will need to revert 558fb26 so "force patch" is used, which saves changes, creates a history record immediately, and doesn't allow for "discard changes" function.

> If the client could patch renditions, should we check the size of thumbnail and include baseImage and viewImage too? Those two still use the old thumbnail Maybe better to do it in superdesk-core?

I'm not sure, but if a video is changed - no matter the size of thumbnails, I think all old thumbnails should be dropped and regenerated. What do you think @ride90? It's even better if we can handle it on the server.

idsvn commented:
> the image that is displayed as a thumbnail NEVER appears in the time range that the thumbnail is representing.

You are right, on your video, previous thumbnail is matched instead of the current one where the cursor is. we will check if we could improve it.
FYI We also have did a quick check with 30s, 1m, 2m and 4m video length, the results are acceptable, they only have a few mismatch

tomaskikutis commented on scripts/extensions/videoEditor/src/VideoTimeline/VideoTimeline.tsx:
```diff
@@ -0,0 +1,294 @@
+
+        let thumbnails = this.props.thumbnails.filter((_, index) => thumbnailsIndex.includes(index));
+
+        // missing some thumbnails on the last second of video, happen when video is too short (<15s)
```
Can you fix this on the back-end so the thumbnaIls aren't missing?

idsvn replied:
We exclude last second because `ffmpeg` (editor tools behind `video-server`) could not capture image on end frames (and 1 second is not much), we are testing with lower value (e.g. 0.05) to see if it's worked and will update this.

tomaskikutis replied:
Even if `ffmpeg` can't generate one thumbnail for example and you want to repeat the last one twice, you should do it on the back-end.

idsvn replied:
That's also an option, we will check which side will be easier and update to you

tomaskikutis commented on scripts/extensions/videoEditor/src/VideoTimeline/VideoTimeline.tsx:
```diff
@@ -0,0 +1,294 @@
+
+        const width = this.props.thumbnails[0].width;
+        // Plus one more incomplete frame because we do not know thumbnails width beforehand to
+        // make timeline render with just enough width space
```
Video aspect ratio & browser screen size are known so thumbnail size could be calculated, no? Even if not, why not delay rendering of the component until the thumbnails are downloaded and sizes are known?

tomaskikutis commented on scripts/extensions/videoEditor/src/VideoEditorThumbnail.tsx:
```diff
@@ -0,0 +1,378 @@
+}
+
+interface IState {
+    type: 'capture' | 'upload' | '';
```
What does empty string mean? It'd be good to drop it if there's no meaning or rename it to a clearer name or to `null` if it means the value is not set.

idsvn replied:
Let's change it to `null`, we need an empty value to check if preview thumbnail is dirty or not.

tomaskikutis commented on scripts/extensions/videoEditor/src/VideoEditorThumbnail.tsx:
```diff
@@ -0,0 +1,378 @@
+
+    componentDidUpdate(prevProps: IProps) {
+        // video has changed so captured thumbnail showed on UI is no longer correct
+        if (prevProps.renditions !== this.props.renditions && this.state.type === 'capture') {
```
```this.state.type === 'capture'```

what is special with the capture state? would it work to instead check if `state.capture != prevState.capture`?

idsvn replied:
No we are not checking `capture` state. We are checking if video has changed or not.
This is when you capture thumbnail but do not click save immediately, but _after editing video_, in that case, current display thumbnail will be differ from the captured. In case upload thumbnail, it doesn't depend on video.

We don't re-draw thumbnail because the video maybe has not finished loading yet, we could wait but reset it is easier

idsvn commented on scripts/extensions/videoEditor/src/VideoTimeline/VideoTimeline.tsx:
```diff
@@ -0,0 +1,296 @@
+        const video = this.props.video;
+        const left = video ? `${(this.state.trim.start / video.duration) * 100}%` : '0%';
+        const right = video ? `${(1 - this.state.trim.end / video.duration) * 100}%` : '0%';
+        const thumbnailWidth = Math.round(video.videoWidth / (video.videoHeight / 50)); // formula from video server
+        // maximum total possible thumbnails, minus margin to avoid making it feels too narrow
+        const totalThumbnails = Math.round(((this.timeline.current?.offsetWidth ?? 0) - 40 * 2) / thumbnailWidth);
+
+        if (!totalThumbnails) {
+            return (<div className={getClass('timeline-controls')} ref={this.timeline} />);
+        }
+
```
Timeline now will be center, so this is to avoid moving left bar when video change (we depend on left bar position in `getPositionInBar` and to calculate which timeline thumbnail to render), kind of place holder

idsvn commented on scripts/extensions/videoEditor/src/VideoEditor.tsx:
```diff
@@ -0,0 +1,788 @@
+                                                etag={this.state.etag}
+                                                renditions={this.state.renditions}
+                                                onToggleLoading={this.handleToggleLoading}
+                                                onSave={(renditions: IArticle['renditions'], etag: string) =>
+                                                    // update preview thumbnail will change etag
+                                                    this.setState({renditions: renditions, etag: etag})
+                                                }
```
Upload thumbnail will change article etag, we think it's possible to do it on the server, but need some hacks, so we think it's better to manually sync it on the client

idsvn commented:
Hi @tomaskikutis , we added comment to clarify some points. 

Regarded missing thumbnail, we updated `video-server` as we have discussed, please check [our branch](https://github.com/idsvn/video-server/tree/safari), also timeline thumbnails now will match the video correctly and no longer display incomplete frame.

Only one problem remain is the preview thumbnail outside (editor) is cached, so it will still display the old one on Firefox (as we commented [here](https://github.com/superdesk/superdesk-client-core/pull/3218/files#diff-043e39fd79fe5d94268dbad05bfb6c9eR222)). For other thumbnail (`viewImage`, `baseImage`), we will update `superdesk-core` once this PR is merged.

tomaskikutis commented on scripts/extensions/videoEditor/src/VideoTimeline/VideoTimeline.tsx:
```diff
@@ -235,11 +225,23 @@ export class VideoTimeline extends React.Component<IProps, IState> {
         const right = video ? `${(1 - this.state.trim.end / video.duration) * 100}%` : '0%';
+        const thumbnailWidth = Math.round(video.videoWidth / (video.videoHeight / 50)); // formula from video server
+        // maximum total possible thumbnails, minus margin to avoid making it feels too narrow
+        const totalThumbnails = Math.round(((this.timeline.current?.offsetWidth ?? 0) - 40 * 2) / thumbnailWidth);
```
Can you avoid duplicating these margin numbers -  `40 * 2`?. Since it doesn't have any connection to the original definition it will render incorrectly if the original number. Instead of duplicating the numbers, could you add a wrapper div? Or at least read the margin from an element so it would stay in sync if the margin was updated?

idsvn replied:
Ah my bad, forgot I could get this via `getComputedStyle`, and this is more like parent `padding` than margin

tomaskikutis commented on scripts/extensions/videoEditor/src/VideoEditor.tsx:
```diff
@@ -39,6 +39,7 @@ interface IState {
     cropEnabled: boolean;
     playing: boolean;
     prevVideoHeight: number;
+    etag: string;
```
It's prone to errors if you store `etag` separately from the entity. Can you store it inside the entity object?

If it's for an article, it should be `article._etag` or if for a video project - `videoProject._etag`.

idsvn replied:
> Can you store it inside the entity object?

Could you clarify more? I removed [article from state](https://github.com/superdesk/superdesk-client-core/pull/3218/commits/c99290faf9443cd301c8b4773d7b7eaf5000baf2) because we only need `renditions` so I'm not sure where to put it, maybe move it out of `state` or to parent?

ride90 replied:
@idsvn I think Tomas said that `_etag` is a part of an article and it should be stored **in** the `article` itself

idsvn replied:
Got it, I will move `renditions` and `etag` to `article` state

idsvn commented on scripts/extensions/videoEditor/src/VideoEditor.tsx:
```diff
@@ -0,0 +1,792 @@
+            {
+                renditions: this.state.article.renditions,
+            },
+            {patchDirectlyAndOverwriteAuthoringValues: true},
```
@tomaskikutis We just noticed that this flag will make updating video thumbnail on _article editor view_ no longer work.
Steps to reproduce:
- Open video in edit mode
- Open Video editor then close it
- Change Video thumbnail and click save

You could see in the `devtools` that although we clicked `save`, no `PATCH` request is made. Logging `AuthoringService` showed that `originalItem` and `item` are the same.

tomaskikutis replied:
Something is wrong with my local video server and I can't start it. Oleg is now working on deploying a test instance with video server enabled and I will take a look at this when the test instance is ready.

tomaskikutis replied:
I just found that there's an issue with `renditions.thumbnail` - you should replace it with `renditions.viewImage` in all places. `thumbnail` rendition is only for backwards-compatibility and will likely be dropped in the future.

> You could see in the devtools that although we clicked save, no PATCH request is made.

Could you record a video with the issue? It might be that because of a large image - if it hasn't finished uploading there will be nothing to save.

Here are a few videos from my testing. There are a few UI glitches when rotating & capturing thumbnails with non-standard sizes. Looks like timeline is missing too.

https://drive.google.com/file/d/1kaQHGAUjbGmg5GdjcuuM4jmr33qX31o2/view?usp=sharing
https://drive.google.com/file/d/1-HMSp2VJHBD7NYozi6GKKYm60TNKrUiO/view?usp=sharing

idsvn replied:
> https://drive.google.com/file/d/1kaQHGAUjbGmg5GdjcuuM4jmr33qX31o2/view?usp=sharing

It seems style is broken, could you check if css is generated and loaded correctly? It should be [like this](https://drive.google.com/file/d/1CnfBdl6kQAaUL3p1e1guwh9XWGj1u9Ub/view?usp=sharing) (1100x500).

> Looks like timeline is missing too

could you send me server error response so I can check it? Just check which request has error with `?action=timeline` in its url is fine.

Regarding thumbnail, after open video editor view and exit, thumbnail is uploaded successfully but save does not apply, you could see on 40s, there should be a PATCH request upon clicking save, here is video: https://drive.google.com/file/d/1vph3ulhmoqXdZA9_wc4-dcIL8m7-heik/view?usp=sharing

ride90 commented:
Hi @idsvn @tomaskikutis , I did some tests on my local and I've noticed a small bug :) 
Here are steps to reproduce it:
- upload video
- upload custom thumbnail [select thumbnail] button (don't open for editing)
- open for editing
- select thumbnail from frame or upload another one
- save and close editing
- ⚠️ thumbnail is not updated

idsvn commented:
> Hi @idsvn @tomaskikutis , I did some tests on my local and I've noticed a small bug :)
> * warning thumbnail is not updated

Hi @ride90, by thumbnail, do you mean [this thumbnail](https://user-images.githubusercontent.com/45254285/90609611-879fef00-e22e-11ea-93cc-9f23c71261d8.png)? It's known issue, because the client has [3 types of thumbnail](https://github.com/superdesk/superdesk-client-core/blob/develop/scripts/core/superdesk-api.d.ts#L356-L358) so we need to remove `viewImage` and `baseImage` when updating thumbnail (see https://github.com/superdesk/superdesk-client-core/pull/3218#issuecomment-650176856)

ride90 commented:
@idsvn , yes, the one you referenced.

ride90 commented:
Hi @idsvn , please fix travis checks and conflicts so we can move forward with this pr.
